### PR TITLE
Make transform application transactional

### PR DESCRIPTION
**Pull request by lava**
### :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [vast.io](https://vast.io), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### CODE REVIEW

File: libvast/test/system/partition.cpp [hunk 1]
```diff
@@ -38,6 +39,16 @@ system::filesystem_actor::behavior_type mock_filesystem(
...
     },
+    [](vast::atom::move, const std::filesystem::path&,
+       const std::filesystem::path&) -> caf::result<vast::atom::done> {
+      vast::die("not implemented");
+    },
+    [](
+      vast::atom::move,
+      const std::vector<std::pair<std::filesystem::path, std::filesystem::path>>&)
+      -> caf::result<vast::atom::done> {
+      vast::die("not implemented");
```
Comment (dominiklohmann) [inline libvast/test/system/partition.cpp:50]:
```suggestion
      FAIL("not implemented");
```

File: libvast/test/system/partition.cpp [hunk 2]
```diff
@@ -38,6 +39,16 @@ system::filesystem_actor::behavior_type mock_filesystem(
     [](atom::read, const std::filesystem::path&) -> caf::result<chunk_ptr> {
       return nullptr;
     },
+    [](vast::atom::move, const std::filesystem::path&,
+       const std::filesystem::path&) -> caf::result<vast::atom::done> {
+      vast::die("not implemented");
```
Comment (dominiklohmann) [inline libvast/test/system/partition.cpp:44]:
```suggestion
      FAIL("not implemented");
```

File: libvast/test/system/partition.cpp [hunk 3]
```diff
@@ -12,6 +12,7 @@
 
 #include "vast/detail/collect.hpp"
 #include "vast/detail/spawn_container_source.hpp"
+#include "vast/die.hpp"
```
Comment (dominiklohmann) [inline libvast/test/system/partition.cpp:15]:
```suggestion
```

File: libvast/include/vast/system/index.hpp [hunk 4]
```diff
@@ -158,20 +158,30 @@ struct index_state {
   [[nodiscard]] std::filesystem::path
   index_filename(const std::filesystem::path& basename = {}) const;
 
+  [[nodiscard]] std::filesystem::path
+  marker_path(const uuid& id) const;
+
   // Maps partitions to their expected location on the file system.
   [[nodiscard]] std::filesystem::path partition_path(const uuid& id) const;
 
+  [[nodiscard]] std::filesystem::path
```
Comment (dominiklohmann) [inline libvast/include/vast/system/index.hpp:167]: Needs docs

File: libvast/include/vast/system/index.hpp [hunk 5]
```diff
@@ -158,20 +158,30 @@ struct index_state {
   [[nodiscard]] std::filesystem::path
   index_filename(const std::filesystem::path& basename = {}) const;
 
+  [[nodiscard]] std::filesystem::path
```
Comment (dominiklohmann) [inline libvast/include/vast/system/index.hpp:161]: Needs docs

File: libvast/fbs/partition_transform.fbs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+include "uuid.fbs";
+
+namespace vast.fbs.partition_transform;
+
+// TODO: Does it make sense to also store the id of
+// the transform itself and the `keep_input` flag?
```
Comment (dominiklohmann) [inline libvast/fbs/partition_transform.fbs:6]: Comment can be removed; makes no sense

File: libvast/fbs/partition_transform.fbs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+table v0 {
+	input_partitions: [UUID];
```
Comment (dominiklohmann) [inline libvast/fbs/partition_transform.fbs:8]: Can you add a comment that this is empty if preserve-input is enabled for compaction?

File: libvast/include/vast/system/filesystem_statistics.hpp [hunk 8]
```diff
@@ -36,12 +36,13 @@ struct filesystem_statistics {
   ops reads;
   ops mmaps;
   ops erases;
+  ops moves;
```
Comment (dominiklohmann) [inline libvast/include/vast/system/filesystem_statistics.hpp:39]: This needs to be added to the status output of the posix filesystem actor

File: libvast/src/system/index.cpp [hunk 9]
```diff
@@ -1579,50 +1759,98 @@ index(index_actor::stateful_pointer<index_state> self,
...
-            }
+            // Record in-progress marker.
+            auto marker_chunk = create_marker(
+              old_partition_ids, new_partition_ids, keep);
+            self
+              ->request(self->state.filesystem, caf::infinite, atom::write_v,
+                        marker_path, marker_chunk)
+              .then(
+                [=, apsv = std::move(apsv)](atom::ok) mutable {
+                  die(":)");
```
Comment (dominiklohmann) [inline libvast/src/system/index.cpp:1770]: :(

**Review by dominiklohmann — APPROVED** [review]
We paired on this and tested it extensively by crashing intentionally in various places.

File: libvast/src/system/posix_filesystem.cpp [hunk 10]
```diff
@@ -23,6 +23,26 @@
 
 namespace vast::system {
 
+caf::expected<atom::done>
+posix_filesystem_state::rename_single_file(const std::filesystem::path& from,
+                                           const std::filesystem::path& to) {
```
Comment (mavam) [inline libvast/src/system/posix_filesystem.cpp:28]: Doesn't rename this more than a single file? Just curious why not go with a plain filesystem rename, which should work for directories as well, no?

Comment (tobim) [reply libvast/src/system/posix_filesystem.cpp:28]: I'm not sure I understand the question. Are you proposing to change the name OR the semantics of this function?

Comment (mavam) [reply libvast/src/system/posix_filesystem.cpp:28]: The name only, because it actually renames directories as well.

Comment (dominiklohmann) [reply libvast/src/system/posix_filesystem.cpp:28]: https://en.wikipedia.org/wiki/Everything_is_a_file

**Pull request merged**